### ISSUE TITLE
Illegal argument in isolate message : (object is a regular Dart Instance)

### ISSUE THREAD

**Issue by joechan-cq** [flutter/flutter#114054]
I want to use another isolate to do the cost-time task. Because the task is depended on some flutter api, so I use `FlutterEngineGroup` to create the default engine(A), and then create another engine(B) also with the same `FlutterEngineGroup`. Then I exchange the `SendPort` of A、B to each other。When I use `SendPort` to send my `TestClz` Object from A to B on **Debug-mode**, it goes wrong：
```
Invalid argument(s): Illegal argument in isolate message : (object is a regular Dart Instance)
```
**Profile** and **Release** works. And It's also ok when send `TestClz` from B to A.

But if I use the `isolate.spawn` instead of FlutterEngineGroup to create new isolate, It works perfectly, both A -> B and B -> A  are ok.

So, I am confused on
1. the difference between Using `FlutterEngineGroup` to create isolate And Using `isolate.spawn`
2. why B -> A is ok, but A -> B not.
3. why profile and release mode are ok

## SouceCode
Override `FlutterActivity` to use `FlutterEngineGroup` to create engines.
``` java
public class MainActivity extends FlutterActivity {

    private static FlutterEngineGroup engineGroup;
    private static FlutterEngine lonerEngine;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        if (engineGroup == null) {
            engineGroup = new FlutterEngineGroup(getApplicationContext());
        }
        super.onCreate(savedInstanceState);
    }

    @Nullable
    @Override
    public FlutterEngine provideFlutterEngine(@NonNull Context context) {
        return engineGroup.createAndRunDefaultEngine(this);
    }

    @Override
    public void configureFlutterEngine(@NonNull FlutterEngine flutterEngine) {
        GeneratedPluginRegistrant.registerWith(flutterEngine);
        if (lonerEngine == null) {
            DartExecutor.DartEntrypoint dartEntrypoint = new DartExecutor.DartEntrypoint(
                    FlutterInjector.instance().flutterLoader().findAppBundlePath(),
                    "lonerMain"
            );
            lonerEngine = engineGroup.createAndRunEngine(getApplicationContext(), dartEntrypoint);
        }
    }
}
```

datastruct to send:  `TestClz`
``` dart
class TestClz {
  final int a;
  final int b;
  TestClz(this.a, this.b);
}
```

`main.dart`
``` dart

@pragma('vm:entry-point')
void lonerMain() {
  SendPort? _loner2mainSendP = IsolateNameServer.lookupPortByName('mainIsolatePort');
  ReceivePort _main2lonerReceiveP = ReceivePort('lonerReceiver');
  _main2lonerReceiveP.listen((msg){
    print('receive msg: $msg');
  });
  _loner2mainSendP?.send(_main2lonerReceiveP.sendPort);
}

void main() {
  runZonedGuarded<Future<void>>(() async {
    WidgetsFlutterBinding.ensureInitialized();

    ReceivePort _main2lonerReceiveP = ReceivePort('mainReceiver');
    _main2lonerReceiveP.listen((msg){
      if (msg is SendPort) {
        SendPort sp = msg;
        try {
          sp.send(TestClz(1, 2));
        } on ArgumentError catch(e) {
          print(e);  // Invalid argument(s): Illegal argument in isolate message : (object is a regular Dart Instance)
        }
      }
    });
    IsolateNameServer.registerPortWithName(
        _main2lonerReceiveP.sendPort, 'mainIsolatePort');

    runApp(MyApp());
  }, (Object error, StackTrace stackTrace) {
//    _reportError(error, stackTrace);
  });
}

```

## Flutter doctor
```
[✓] Flutter (Channel stable, 2.5.3, on macOS 11.5.2 20G95 darwin-arm, locale zh-Hans-CN)
    • Flutter version 2.5.3 at /Users/xxx/Desktop/SDK/fvm/2.5.3
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 18116933e7 (1 year ago), 2021-10-15 10:46:35 -0700
    • Engine revision d3ea636dc5
    • Dart version 2.14.4
    • Pub download mirror https://pub.flutter-io.cn
    • Flutter download mirror https://storage.flutter-io.cn

[✓] Android toolchain - develop for Android devices (Android SDK version 32.0.0)
    • Android SDK at /Users/xxx/Desktop/SDK/android-sdk-macosx
    • Platform android-32, build-tools 32.0.0
    • ANDROID_HOME = /Users/xxx/Desktop/SDK/android-sdk-macosx
    • ANDROID_SDK_ROOT = /Users/xxx/Desktop/SDK/android-sdk-macosx
    • Java binary at: /Users/xxx/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/213.7172.25.2113.9123335/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 13.2.1, Build version 13C100
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.3)
    • Android Studio at /Users/xxx/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/213.7172.25.2113.9123335/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

```

**Comment by navaronbracke** [comment]
You can only send specific types through a `SendPort`. Consider converting your class to a Map and sending that instead.

https://api.dart.dev/stable/2.18.3/dart-isolate/SendPort/send.html

**Comment by exaby73** [comment]
Hello @joechan-cq. I noticed you're using an older version of Flutter from your doctor output (2.5.3). Could you confirm if this issue is still prevalent in that latest stable (3.3.5)?

**Comment by joechan-cq** [comment]
> Hello @joechan-cq. I noticed you're using an older version of Flutter from your doctor output (2.5.3). Could you confirm if this issue is still prevalent in that latest stable (3.3.5)?

I try it on flutter 3.3.6, it works.

If possible, may i ask for the reason?

@exaby73

**Comment by exaby73** [comment]
@joechan-cq Glad it works :) I couldn't tell you exactly what fixed this issue as there has been a lot of PRs merged since 2.5.3
